packages/firestore/src/remote/persistent_stream.ts

```diff
@@ -253,6 +245,108 @@ export abstract class PersistentStream<
   }
 
   /**
+   * Initializes the idle timer. If no write takes place within one minute, the
+   * WebChannel stream will be closed.
+   */
+  markIdle(): void {
+    this.idle = true;
+    this.queue
+      .schedule(() => {
+        return this.handleIdleCloseTimer();
+      }, IDLE_TIMEOUT_MS)
+      .catch((err: FirestoreError) => {
+        // We ignore Promise rejections for cancelled idle checks.
+        assert(
+          err.code === Code.CANCELLED,
+          `Received unexpected error in idle timeout closure. Expected CANCELLED, but was: ${err}`
+        );
+      });
+  }
+
+  /** Sends a message to the underlying stream. */
+  protected sendRequest(msg: SendType): void {
+    this.cancelIdleCheck();
+    this.stream!.send(msg);
+  }
+
+  /** Called by the idle timer when the stream should close due to inactivity. */
+  private handleIdleCloseTimer(): Promise<void> {
+    if (this.isOpen() && this.idle) {
+      // When timing out an idle stream there's no reason to force the stream into backoff when
+      // it restarts so set the stream state to Initial instead of Error.
+      return this.close(PersistentStreamState.Initial);
+    }
+    return Promise.resolve();
+  }
+
+  /** Marks the stream as active again. */
+  private cancelIdleCheck() {
+    this.idle = false;
+  }
+
+  /**
+   * Closes the stream and cleans up as necessary:
+   *
+   * <ul>
+   *   <li>closes the underlying GRPC stream;
+   *   <li>calls the onClose handler with the given 'error';
+   *   <li>sets internal stream state to 'finalState';
+   *   <li>adjusts the backoff timer based on the error
+   * </ul>
+   *
+   * A new stream can be opened by calling `start` unless `finalState` is set to
+   * `PersistentStreamState.Stopped`.
+   *
+   * @param finalState the intended state of the stream after closing.
+   * @param error the error the connection was closed with.
+   */
+  private close(
+    finalState: PersistentStreamState,
+    error?: FirestoreError
+  ): Promise<void> {
+    assert(
+      finalState == PersistentStreamState.Error || isNullOrUndefined(error),
```

Comment:
As in the iOS PR, this allows the error to be nil if the finalState is FSTStreamStateError. Is that desirable?

Reply:
This gets invoked by 'handleStreamClose' (the GRPC callback handler) and making this stricter would require that the server only ever closes a stream with an error.